### PR TITLE
Feature/buf client

### PR DESCRIPTION
**Pull request by jdechicchis**
Adding support for protobuf decoding using BSR

### CODE REVIEW

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java [hunk 1]
```diff
@@ -1,7 +1,142 @@
...
+
+    for (int i = 0; i < fileDescriptorSet.getFileCount(); i++) {
+      FileDescriptorProto p = fileDescriptorSet.getFile(i);
+      descriptorProtoIndexBuilder.put(p.getName(), p);
+    }
+
+    ImmutableMap<String, FileDescriptorProto> descriptorProtoIndex = descriptorProtoIndexBuilder.build();
+    Map<String, FileDescriptor> descriptorCache = new HashMap<String, FileDescriptor>();
+
+    HashMap<String, FileDescriptor> allFileDescriptors = new HashMap<String, FileDescriptor>();
```
Comment (csueiras) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java]: consider using `Map<String, FileDescriptor>` as your type on the left.

Also no need to specify the type in the `new HashMap<...>` instead use the diamond operator:

```java
final Map<String, FileDescriptor> allFileDescriptors = new HashMap<>();
```

You should do this everywhere where possible in this class

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java [hunk 2]
```diff
@@ -1,7 +1,142 @@
...
+    try {
+      for (int i = 0; i < fileDescriptorSet.getFileCount(); i++) {
+        FileDescriptor desc = descriptorFromProto(fileDescriptorSet.getFile(i), descriptorProtoIndex, descriptorCache);
+        allFileDescriptors.put(desc.getName(), desc);
+      }
+    } catch (DescriptorValidationException e) {
+      log.error("Failed to create dependencies map {}", e);
+    }
+
+    for (FileDescriptor fileDesc : allFileDescriptors.values()) {
```
Comment (csueiras) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java]: you can do this functionally too (just sharing, in this particular I dont think it would make this code any better)

```java
allFileDescriptors.values()
  .stream()
  .filter(f => f.getPackage() == packageName)
  .map(f => f.findMessageTypeByName(typeName))
  .filter(Objects::nonNull)
  .findFirst()
  .orElse(null)
```

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java [hunk 3]
```diff
@@ -1,7 +1,142 @@
...
+        if (desc != null) {
+          return desc;
+        }
+      }
+    }
+
+    return null;
+  }
+
+  private Image getImage(String owner, String repo) throws StatusRuntimeException {
```
Comment (csueiras) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java:104]: consider inlining, would likely make this a bit nicer to read:

```java
return bufClient.getImage(GetImageRequest.newBuilder()
    .setOwner(owner)
    .setRepository(repo)
    .setReference("main")
    .build())
    .getImage();
```

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java [hunk 4]
```diff
@@ -1,7 +1,142 @@
...
+        .build();
+    GetImageResponse response = bufClient.getImage(request);
+    return response.getImage();
+  }
+
+  // From
+  // https://github.com/grpc-swagger/grpc-swagger/blob/master/grpc-swagger-core/src/main/java/io/grpc/grpcswagger/grpc/ServiceResolver.java#L118.
+  private FileDescriptor descriptorFromProto(
+      FileDescriptorProto descriptorProto,
+      ImmutableMap<String, FileDescriptorProto> descriptorProtoIndex,
```
Comment (csueiras) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java]: use interface `Map<String, FileDescriptorProto>`

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java [hunk 5]
```diff
@@ -1,7 +1,142 @@
...
+      Map<String, FileDescriptor> descriptorCache) throws DescriptorValidationException {
+    // First, check the cache.
+    String descriptorName = descriptorProto.getName();
+    if (descriptorCache.containsKey(descriptorName)) {
+      return descriptorCache.get(descriptorName);
+    }
+
+    // Then, fetch all the required dependencies recursively.
+    ImmutableList.Builder<FileDescriptor> dependencies = ImmutableList.builder();
```
Comment (csueiras) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java]: there's basically no real benefit to using immutable collections in 100% private API, consider using an `ArrayList` instead

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java [hunk 6]
```diff
@@ -1,7 +1,142 @@
...
+    headers.put(bearerKey, String.format("Bearer %s", apiKey));
+
+    bufClient = bufClient.withInterceptors(MetadataUtils.newAttachHeadersInterceptor(headers));
+  }
+
+  @Nullable
+  public Descriptor getDescriptor(String owner, String repo, String fullyQualifiedTypeName) {
+    List<String> parts = Arrays.asList(fullyQualifiedTypeName.split("\\."));
+
+    if (parts.size() == 0) {
```
Comment (csueiras) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java]: its better practice for this type of check to use `isEmpty()` instead

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java [hunk 7]
```diff
@@ -1,7 +1,142 @@
...
+
+    Metadata headers = new Metadata();
+    Metadata.Key<String> bearerKey = Metadata.Key.of("authorization", Metadata.ASCII_STRING_MARSHALLER);
+    headers.put(bearerKey, String.format("Bearer %s", apiKey));
+
+    bufClient = bufClient.withInterceptors(MetadataUtils.newAttachHeadersInterceptor(headers));
+  }
+
+  @Nullable
+  public Descriptor getDescriptor(String owner, String repo, String fullyQualifiedTypeName) {
```
Comment (csueiras) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java]: i would kill all usage of the immutable collections, there's no need for them, you get no benefits but all of the annoyances of it (having to "build" the outcome collection)

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java [hunk 8]
```diff
@@ -20,104 +35,319 @@
...
-    // extract fqtn
+    // Extract the fully qualified type name from the topic.
+    // E.g., my-topic.proto.service.v1.User -> service.v1.User.
     String[] parts = topic.split("\\.proto\\.");
     ProtoSchema ret = new ProtoSchema();
     ret.setFullyQualifiedTypeName(parts[parts.length - 1]);
     return ret;
   }
 
+  private DeserializedKeyValue deserializeProtobuf(ConsumerRecord<Bytes, Bytes> msg,
```
Comment (csueiras) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java:218]: any time you do `new String(someBytes)` and you don't specify the encoding, you are at the mercy of your OS default encoding. Use `new String(someBytes, StandardCharsets.UTF_8)`

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java [hunk 9]
```diff
@@ -20,104 +35,319 @@
...
+          builder.value(parse(msg.value().get(), valueDescriptor));
+          builder.valueFormat(MessageFormat.PROTOBUF);
+        } else {
+          builder.value(new String(msg.value().get()));
+          builder.valueFormat(MessageFormat.UNKNOWN);
+        }
+      }
+      return builder.build();
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to parse record from topic " + msg.topic(), e);
```
Comment (csueiras) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java]: use `new UncheckedIOException(e)`

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java [hunk 10]
```diff
@@ -20,104 +35,319 @@
...
+      @Nullable String key,
+      @Nullable String data,
+      @Nullable Integer partition) {
+    if (data == null) {
+      return new ProducerRecord<>(topic, partition, Objects.requireNonNull(key).getBytes(), null);
+    }
+
+    ProtoSchema protoSchema = protoSchemaFromTopic(topic);
+    if (protoSchema == null) {
+      throw new RuntimeException("Could not infer proto schema from topic " + topic);
```
Comment (csueiras) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java]: RuntimeException is a pretty crappy exception, you should try to use a more specific exception. In this case I would suggest using `IllegalStateException`

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java [hunk 11]
```diff
@@ -20,104 +35,319 @@
...
+    }
+
+    DynamicMessage.Builder builder = DynamicMessage.newBuilder(descriptor);
+
+    DynamicMessage message;
+    try {
+      JsonFormat.parser().merge(data, builder);
+      message = builder.build();
+    } catch (InvalidProtocolBufferException e) {
+      throw new RuntimeException("Failed to merge record for topic " + topic, e);
```
Comment (csueiras) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java]: I bet `InvalidProtocolBufferException` is an IOException sub class, so you can just use `new UncheckedIOException(e)`

**Review by csueiras — CHANGES_REQUESTED** [review]
Various changes requested

File: .github/clear-street/README.md [hunk 12]
```diff
@@ -12,9 +16,17 @@ Use the [Protobuf on Kafka](https://www.notion.so/clearstreet/Protobuf-on-Kafka-
 Use these settings to set up the Buf connection:
 
 ```yaml
-bufRegistry: https://<url>
-bufUsername: <user>
+bufRegistry: <url>
+bufPort: <port>
 bufApiToken: <token>
+bufRegistryCacheDurtionSeconds: <seconds>
```
Comment (ottaviohartman) [inline .github/clear-street/README.md]:
```suggestion
bufRegistryCacheDurationSeconds: <seconds>
```

Comment (ottaviohartman) [inline .github/clear-street/README.md]: We've avoided unpacking these YAML's, but since there are so many `buf*` keys it may make sense to add a structured `buf:` section. Can do this in the future, too.

File: .gitignore [hunk 13]
```diff
@@ -1,5 +1,6 @@
 HELP.md
 target/
+/kafka-ui-proto-gen/src/
```
Comment (ottaviohartman) [inline .gitignore]: nit: Add this to the `.gitignore` in `kafka-ui-proto-gen`. When possible, we should try not to edit their files

File: kafka-ui-api/Dockerfile [hunk 14]
```diff
@@ -1,7 +1,6 @@
-FROM alpine:3.15.0
+FROM openjdk:13.0.2-jdk-buster
```
Comment (ottaviohartman) [inline kafka-ui-api/Dockerfile]: Why is this necessary?

Can we create a new Dockerfile so that we don't have conflicts with upstream master if they change this file?

File: kafka-ui-api/pom.xml [hunk 15]
```diff
@@ -1,6 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<project xmlns="http://maven.apache.org/POM/4.0.0"
-    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
Comment (ottaviohartman) [inline kafka-ui-api/pom.xml:2]: Why changes here? Is there a way to not have this changes in this file?

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java [hunk 16]
```diff
@@ -1,7 +1,142 @@
...
+    bufClient = ImageServiceGrpc.newBlockingStub(channel);
+
+    Metadata headers = new Metadata();
+    Metadata.Key<String> bearerKey = Metadata.Key.of("authorization", Metadata.ASCII_STRING_MARSHALLER);
+    headers.put(bearerKey, String.format("Bearer %s", apiKey));
+
+    bufClient = bufClient.withInterceptors(MetadataUtils.newAttachHeadersInterceptor(headers));
+  }
+
+  @Nullable
```
Comment (ottaviohartman) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java]: This may be wrong, but I think @Nullable should come before `Descriptor`?

Comment (ottaviohartman) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java]: Oh, nevermind: https://stackoverflow.com/questions/25958995/where-to-put-nullable-on-methods-with-nullable-return-types

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java [hunk 17]
```diff
@@ -20,104 +35,319 @@
...
     } else {
-      // used for testing
+      // Used for testing.
       this.schemaRegistryAwareRecordSerDe = new SchemaRegistryAwareRecordSerDe(cluster, schemaRegistryClient);
     }
-    this.cluster = cluster;
+
     this.bufClient = bufClient;
+
+    if (cluster.getBufRegistryCacheDurtionSeconds() != null) {
```
Comment (ottaviohartman) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java]: Durtion -> Duration

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java [hunk 18]
```diff
@@ -20,104 +35,319 @@
...
+      this.cachedMessageDescriptorRetensionSeconds = cluster.getBufRegistryCacheDurtionSeconds();
```
Comment (ottaviohartman) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java]: Retension -> Retention

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java [hunk 19]
```diff
@@ -20,104 +35,319 @@
...
+
+    return this.deserializeProtobuf(msg, keyDescriptor, valueDescriptor);
+  }
+
+  @Nullable
+  ProtoSchema protoKeySchemaFromHeaders(Headers headers) {
+    // Get PROTOBUF_TYPE_KEY header.
     String fullyQualifiedTypeName = null;
-    for (Header header : headers.headers("PROTOBUF_TYPE")) {
+    for (Header header : headers.headers("PROTOBUF_TYPE_KEY")) {
```
Comment (ottaviohartman) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java:176]: PROTOBUF_TYPE_KEY? This wasn't updated in the docs https://www.notion.so/clearstreet/Protobuf-on-Kafka-ac912401cf504c75956d907b86a12f55

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java [hunk 20]
```diff
@@ -20,104 +35,319 @@
...
+    ProtoSchema ret = new ProtoSchema();
+    ret.setFullyQualifiedTypeName(fullyQualifiedTypeName);
+    return ret;
+  }
+
+  @Nullable
+  ProtoSchema protoValueSchemaFromHeaders(Headers headers) {
+    // Get PROTOBUF_TYPE_VALUE header.
+    String fullyQualifiedTypeName = null;
+    for (Header header : headers.headers("PROTOBUF_TYPE_VALUE")) {
```
Comment (ottaviohartman) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java:192]: What happened to SCHEMA_ID?

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java [hunk 21]
```diff
@@ -20,104 +35,319 @@
...
+        }
+      }
+      return builder.build();
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to parse record from topic " + msg.topic(), e);
+    }
+  }
+
+  private static long getDateDiffMinutes(Date date1, Date date2, TimeUnit timeUnit) {
+    long diffInMillies = date2.getTime() - date1.getTime();
```
Comment (ottaviohartman) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java]: nit: diffInMillis

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java [hunk 22]
```diff
@@ -20,104 +35,319 @@
...
+    return timeUnit.convert(diffInMillies, TimeUnit.MILLISECONDS);
+  }
+
+  @Nullable
+  private Descriptor getDescriptor(String fullyQualifiedTypeName) {
+    Date currentDate = new Date();
+    CachedDescriptor cachedDescriptor = cachedMssageDescriptorMap.get(fullyQualifiedTypeName);
```
Comment (ottaviohartman) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java]: Mssage -> Message

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java (hunk 20 above)
Comment (jdechicchis) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java:192]: I thought it may be better to remove it for now since we don't support it / do anything with it. If we add support in the future would be fairly easy to add back

File: kafka-ui-api/Dockerfile (hunk 14 above)
Comment (jdechicchis) [reply kafka-ui-api/Dockerfile]: We need this because the Java grpc package relies on stuff that alpine is missing. All make a new clst.Dockerfile though

File: kafka-ui-api/pom.xml (hunk 15 above)
Comment (jdechicchis) [reply kafka-ui-api/pom.xml:2]: We need the addition of this dep so tings build properly
```
<dependency>
            <groupId>com.provectus</groupId>
            <artifactId>kafka-ui-proto-gen</artifactId>
            <version>${project.version}</version>
        </dependency>
```

File: .github/clear-street/README.md (hunk 12 above)
Comment (jdechicchis) [reply .github/clear-street/README.md]: Agreed, I'll ticket this out and tackle in a subsequent PR

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java (hunk 19 above)
Comment (jdechicchis) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java:176]: Updated the doc

**Review by ottaviohartman — APPROVED** [review]

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java [hunk 23]
```diff
@@ -1,7 +1,140 @@
...
```
Comment (csueiras) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/client/BufSchemaRegistryClient.java]: since you aren't implementing an interface here... you could do the better thing and return `Optional<Descriptor>` instead. Nullable is a bad crutch

**Review by csueiras — APPROVED** [review]
